VSRAD.DebugServer/ObservableProcess.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+using VSRAD.DebugServer.IPC.Responses;
+
+namespace VSRAD.DebugServer
+{
+    public sealed class ObservableProcess
+    {
+        public event EventHandler ExecutionStarted;
+        public event EventHandler<string> StdoutRead;
+        public event EventHandler<string> StderrRead;
+
+        private readonly Process _process;
+        private readonly bool _waitForCompletion;
+        private readonly int _timeout;
+        private bool _stoppedByTimeout = false;
+
+        public ObservableProcess(IPC.Commands.Execute command)
+        {
+            _process = new Process
+            {
+                StartInfo = new ProcessStartInfo(command.Executable, command.Arguments)
+                {
+                    WorkingDirectory = command.WorkingDirectory
+                },
+                EnableRaisingEvents = true
+            };
+            if (command.RunAsAdministrator)
+            {
+                _process.StartInfo.UseShellExecute = true;
+                _process.StartInfo.Verb = "runas";
+            }
+            else
+            {
+                _process.StartInfo.UseShellExecute = false;
+                _process.StartInfo.RedirectStandardOutput = true;
+                _process.StartInfo.RedirectStandardError = true;
+            }
+            _waitForCompletion = command.WaitForCompletion;
+            _timeout = command.ExecutionTimeoutSecs;
+        }
+
+        public async Task<ExecutionCompleted> StartAndObserveAsync()
+        {
+            if (!_waitForCompletion)
+                return RunWithoutAwaitingCompletion();
+
+            var processExitedTcs = new TaskCompletionSource<bool>();
+            _process.Exited += (sender, args) => processExitedTcs.SetResult(true);
+
+            var (stdoutTask, stderrTask) = InitializeOutputCapture();
+
+            var stopWatch = Stopwatch.StartNew();
+            try
+            {
+                _process.Start();
+                ExecutionStarted?.Invoke(this, new EventArgs());
+                if (_process.StartInfo.RedirectStandardOutput)
+                {
+                    _process.BeginOutputReadLine();
+                    _process.BeginErrorReadLine();
+                }
+                SetProcessTimeout();
+            }
+            catch (System.ComponentModel.Win32Exception)
+            {
+                return new ExecutionCompleted { Status = ExecutionStatus.CouldNotLaunch };
+            }
+            catch (InvalidOperationException) // missing executable name
+            {
+                return new ExecutionCompleted { Status = ExecutionStatus.CouldNotLaunch };
+            }
+
+            await processExitedTcs.Task;
+
+            var status = _stoppedByTimeout ? ExecutionStatus.TimedOut : ExecutionStatus.Completed;
+
+            var stdout = await stdoutTask;
+            var stderr = await stderrTask;
```

Comment:
I don't see where you finish this tasks. Are you sure you need the tasks here at all? You already have a tcs you create checking on process exit. I guess you don't need tasks for stdout/stderr at all here, just stringbuilders.

Reply:
The tasks are completed when `OutputDataReceived` and `ErrorDataReceived` events fire with `null`, which indicates
that there's no more output. This can happen after the `Exited` event due to output buffering.

If I understood your suggestion correctly, we should return `StringBuilder`s from `InitializeOutputCapture`
and read them after `await processExitedTcs.Task;`. This may lead to a loss of data for large outputs, as demonstrated by [this patch](https://gist.github.com/timlathy/7eda75dd19d547e0c140f80d4ea94440) (the test fails with modifications and passes without them, i.e. with `await`).